pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py

```diff
@@ -218,6 +223,45 @@ def _request_generator_thread(self):
             _LOGGER.debug('Sending request:\n%r', request)
             yield request
 
+    def _stop_request_generator(self, request_generator):
+        """Ensure a request generator is closed.
+
+        This **must** be done when recovering from a retry-able exception.
+        If not, then an inactive request generator (i.e. not attached to any
+        actual RPC) will be trying to access the same request queue as the
+        active request generator.
+
+        In addition, we want the gRPC thread consuming to cleanly exit so
+        that system resources are not wasted.
+
+        Args:
+            request_generator (Generator): A streaming pull request generator
+                returned from :meth:`_request_generator_thread`.
+
+        Raises:
+            ValueError: If ``.close()`` fails for any reason other than
+                "generator already executing".
+            ValueError: If the queue is not empty and the generator is
+                running.
+        """
+        with self._put_lock:
+            try:
+                request_generator.close()
+            except ValueError as exc:
+                if exc.args != ('generator already executing',):
+                    raise
+                if not self._request_queue.empty():
+                    raise ValueError('Queue expected to be empty.')
+                # If we **cannot** close the request generator,
+                # then there is no blocking get on the queue. Since
+                # we have locked ``.put()`` this means that the
+                # queue **was** and remains empty.
+                self._request_queue.put(_helper_threads.STOP)
+                # Wait for the request generator to ``.get()`` the ``STOP``.
+                while not self._request_queue.empty():
```

Comment:
if we use `task_done` we can use [queue.join](https://docs.python.org/2/library/queue.html#Queue.Queue.join).